Create & document steps for making release builds

kellyi opened the pull request:
## Overview

This PR adjusts the project scripts & readme to set the project up for making release build via Jenkins. In addition to adding a `cibuild` script to build the assets, the changes here also adjust the `setup` script to run `npm install` conditionally depending on whether `PWD_RIVER_DISPATCHES_RELEASE` has been set as an environment variable. If that var is unset, we run `npm install`; if it is, then we don't.

Along with the changes here I've also set up a release build job on new Jenkins which runs `setup` and `cibuild` with the env var set. The project assets currently live in `/var/local/pwd-river-dispatches-assets` there, and the build job configuration uses that value as the `setup` argument.

That Jenkins job builds a zipfile which includes the relevant changes from this repo, a version.txt file, and the assets. I currently have the job set to build this branch, but I can set it to build `release*` when this is in.

Connects #3 
Connects #11

## Testing

* visit the pwd-river-dispatches-release job on new Jenkins and build it. Verify that the project builds successfully
* get the zip file from that build and load the Dispatches...html file in the browser. Verify that the project works as it should and that the assets are inlcuded
* rerun `setup` on your local and verify that it *does* run `npm install`

Commit:
Create cibuild script

- add cibuild script for Jenkins build job
- add dist dir for output
- update .gitignore

Commit:
Add release instructions

- add git-flow release instructions to README.md
- add CHANGELOG.md

Commit:
Add env check to setup script

- only run `npm install` if PWD_RIVER_DISPATCHES_RELEASE env var unset
- replace tabs with spaces :art:

kellyi commented:
May also be worth taking a look at the configuration for the Jenkins build job. Since the full zip file is so large, I've set it to keep only the last 5 builds around, but we could reduce this to 2 maybe?

kellyi commented:
Picked up #8 as a continuation of this.

rajadain commented on scripts/setup:
```diff
@@ -5,11 +5,13 @@ set -ex
-	echo "Could not find source directory $SOURCE_DIR"
-	exit 1
+    echo "Could not find source directory $SOURCE_DIR"
+    exit 1
```
Sorry about the tabs, this was probably from Sublime Text which defaults to tabs. I was using Sublime because only that could open the large Dispatches...js file.

kellyi replied:
No worries!

rajadain commented on .gitignore:
```diff
@@ -1,3 +1,8 @@
+version.txt
+dist/*
+!dist/.gitkeep
+
```
Empty line

rajadain commented on scripts/cibuild:
```diff
@@ -0,0 +1,22 @@
+        usage
+    else
+
+git rev-parse HEAD > version.txt
```
Why the odd indentation?

kellyi replied:
This was here because I was imitating a similar script for AKRF; I'll fix the indentation here.

rajadain commented:
The `pwd-river-dispatches-release` job isn't part of the PWD group on Jenkins: http://civicci01.internal.azavea.com/view/pwd/ . Can we make it so?

rajadain commented:
There is a post-build task of `vagrant halt` that is causing errors on Jenkins:

```
Performing Post build task...
Match found for : : True
Logical operation result is TRUE
Running script  : vagrant halt
[pwd-river-dispatches-release] $ /bin/sh -xe /tmp/jenkins2317241564545317854.sh
+ vagrant halt
A Vagrant environment or target machine is required to run this
command. Run `vagrant init` to create a new Vagrant environment. Or,
get an ID of a target machine from `vagrant global-status` to run
this command on. A final option is to change to a directory with a
Vagrantfile and to try again.
POST BUILD TASK : FAILURE
END OF POST BUILD TASK : 0
```

Since this project does not use Vagrant, we should remove that.

rajadain reviewed (changes requested):
Instead of the CI running two scripts:

![image](https://user-images.githubusercontent.com/1430060/27668299-52842e6c-5c4e-11e7-876f-824e7e82180a.png)

It should only ever run one: `cibuild`.

Consider renaming the current implementation of `cibuild` to `cipublish`, and make a new `cibuild` that calls `setup` and `cipublish` in sequence. The location of the original package `/var/local/pwd-river-dispatches-assets` may be provided either as argument to `cibuild`, or as an environment variable. Both are acceptable.

kellyi commented:
> Since this project does not use Vagrant, we should remove that.

Done!

Commit:
Split cibuild and cipublish scripts

- create separate cipublish script to handle creating the release zip
file
- adjust cibuild to build the project on Jenkins

rajadain commented on scripts/cipublish:
```diff
@@ -0,0 +1,22 @@
+
+    git rev-parse HEAD > version.txt
+    zip -r dist/pwd-river-dispatches-release.zip usgs-details.html version.txt \
+    images/ CHANGELOG.md media/ edge_includes/ Dispatches_v04t* -x *.zip
```
Can we indent this line one level deeper to communicate easier that it is an extension of the previous one?

rajadain commented on scripts/cibuild:
```diff
@@ -0,0 +1,20 @@
+    else
+
+    ./scripts/setup "$1"
+
```
Shouldn't `cibuild` call `cipublish`?

kellyi replied:
Right now the Jenkins job runs `cibuild` and `cipublish` in succession. I checked a few jobs and it looks that roughly matches what the others do. The `setup` thing was another imitation of ARKF: looks like we call `setup` there then `cibuild`.

rajadain replied:
Alright let's keep that then.

rajadain reviewed (approved):
Looks good, great job on this!

rajadain commented:
Friendly reminder to squash before merging.

kellyi commented:
Thanks! I checked and this is building correctly on Jenkins now. Going to tweak the job a little more to get it beneath the `pwd` jobs, and then I'll squash and merge this -- and make an initial release just to check that that part works, too.

kellyi commented:
Thanks for your help with this!

kellyi commented:
Made this release using the steps in the README https://github.com/azavea/pwd-river-dispatches/releases/tag/0.0.1 after merging, and pointed Jenkins to build master.
